### PR TITLE
Refactor: Keyboard Repetition

### LINKED ISSUE #356: Keyboard repeat sometimes blasts
From time to time I've noticed that I get a key repeated multiple times without the key being held for long enough, just arbitrary happens while typing. Maybe the timer invalidation is crewed in some cases?

I don't have clear steps to reproduce other than use alacritty with wayland-rs 0.30 for a day and see things break from time to time...

**This doesn't happen with sctk 0.16.**

### PR DESCRIPTION
**Pull request by wash2**
Refactors keyboard repetition closer to original 0.16 implementation and fixes #356

### CODE REVIEW

**Comment by wash2** [comment]
Even if I use an arc mutex to share the repeat key, instead of using a channel to communicate repeat events, as in my last commit, there continue to be spurious repeats. It seems to me that the `Released` event is just not processed until after the timer event goes off, even though it occurred first. I'm not super sure where to go from here, but I'll keep thinking on it :sweat_smile:

**Comment by kchibisov** [comment]
Could we simply remove the timer source and re-arm on press? I think dropping timer source should be faster, no?

For me the events are quite seldom though. It's also strange that `Arc<Mutex<>>` approach isn't working, maybe something is not communicated in time?

**Comment by wash2** [comment]
> Could we simply remove the timer source and re-arm on press? I think dropping timer source should be faster, no?

I'll try that, but I think if I understand what you mean, it should have the same-ish effect that using `Arc<Mutex<>>` does.

**Comment by wash2** [comment]
Ok and I guess a small clarification is, I can't produce a repeat before a release a repeat now even with just 60ms delay after spamming a single key for a while. I can produce a repeat when quickly pressing multiple keys though.

**Comment by kchibisov** [comment]
I guess that's my issue actually, I do press multiple keys near at the same time given that I type fast. I haven't tested any of your recent patches though.

**Comment by wash2** [comment]
I've refactored the key repetition to be more like how it was implemented in 0.16, but with a delay of 60, I can still produce spurious repeats when key-mashing. I don't think the timer is firing at the wrong time, but what seems to happen is that it sometimes blocks until it goes off, preventing other key events from being processed in the meantime?

That being said, I think for now I'll mark this ready for review again :sweat_smile: I think the refactor is still probably an improvement.

**Comment by wash2** [comment]
Ok, removing the ping source seems to fix the spurious repeats. Sorry for spamming so much with this PR.

**Comment by kchibisov** [comment]
>Ok, removing the ping source seems to fix the spurious repeats. Sorry for spamming so much with this PR.

@wash2 I'll add it into my local bind then for testing and report back, thx.

File: src/seat/keyboard/mod.rs [hunk 1]
```diff
@@ -657,13 +670,54 @@ where
...
+                                                    }
+                                                    let key = key.as_mut().unwrap();
+                                                    key.key.time += if key.is_first {
+                                                        key.is_first = false;
+                                                        delay
+                                                    } else {
+                                                        gap.as_millis() as u32
+                                                    };
+                                                    let e = key.key.clone();
+                                                    callback(e, t);
```
[the remark this answers is not in the document]
Comment (wash2) [reply src/seat/keyboard/mod.rs:715]: Yup!

**Review by kchibisov — CHANGES_REQUESTED** [review]

**Review by kchibisov — APPROVED** [review]
Haven't seen any issue while using alacritty for 2 days. So I guess the issue is resolved.

File: src/seat/keyboard/repeat.rs [hunk 2]
```diff
@@ -17,32 +15,28 @@ use super::{
...
-    /// Key has changed during the repeat, but the repeat shouldn't stop.
-    KeyChanged(KeyEvent),
+pub type RepeatCallback<T> = Box<dyn FnMut(&mut T, &WlKeyboard, KeyEvent) + 'static>;
 
-    /// The repeat info has changed.
-    RepeatInfo(RepeatInfo),
+pub(crate) struct RepeatData<T> {
+    pub(crate) current_repeat: Option<RepeatedKey>,
+    pub(crate) repeat_info: RepeatInfo,
+    pub(crate) loop_handle: LoopHandle<'static, T>,
```
Comment (i509VCB) [inline src/seat/keyboard/repeat.rs:29]: I think this will fail to compile if the `calloop` feature is disabled.

Comment (wash2) [reply src/seat/keyboard/repeat.rs:29]: ya, but the repeat module is feature gated. I did check just now though and it seems another feature gate is missing.

**Review by i509VCB — APPROVED** [review]
Seems fine to me at least.

**Pull request merged**